sql/core/src/test/scala/org/apache/spark/sql/DataFrameSetOperationsSuite.scala

```diff
@@ -743,17 +777,59 @@ class DataFrameSetOperationsSuite extends QueryTest with SharedSparkSession {
       StructField("a", StringType)))
     val nestedStructValues2 = Row("b", "a")
 
-    val df1: DataFrame = spark.createDataFrame(
+    val df1 = spark.createDataFrame(
       sparkContext.parallelize(Row(nestedStructValues1) :: Nil),
       StructType(Seq(StructField("topLevelCol", nestedStructType1))))
 
-    val df2: DataFrame = spark.createDataFrame(
+    val df2 = spark.createDataFrame(
       sparkContext.parallelize(Row(nestedStructValues2) :: Nil),
       StructType(Seq(StructField("topLevelCol", nestedStructType2))))
 
     val union = df1.unionByName(df2, allowMissingColumns = true)
-    checkAnswer(union, Row(Row(null, "b")) :: Row(Row("a", "b")) :: Nil)
-    assert(union.schema.toDDL == "`topLevelCol` STRUCT<`a`: STRING, `b`: STRING>")
+    assert(union.schema.toDDL == "`topLevelCol` STRUCT<`b`: STRING, `a`: STRING>")
+    checkAnswer(union, Row(Row("b", null)) :: Row(Row("b", "a")) :: Nil)
+  }
+
+  test("SPARK-35290: Make unionByName null-filling behavior work with struct columns"
+      + " - sorting edge case") {
+    val nestedStructType1 = StructType(Seq(
+      StructField("b", StructType(Seq(
+        StructField("ba", StringType)
+      )))
+    ))
+    val nestedStructValues1 = Row(Row("ba"))
+
+    val nestedStructType2 = StructType(Seq(
+      StructField("a", StructType(Seq(
+        StructField("aa", StringType)
+      ))),
+      StructField("b", StructType(Seq(
+        StructField("bb", StringType)
+      )))
+    ))
+    val nestedStructValues2 = Row(Row("aa"), Row("bb"))
+
+    val df1 = spark.createDataFrame(
+      sparkContext.parallelize(Row(nestedStructValues1) :: Nil),
+      StructType(Seq(StructField("topLevelCol", nestedStructType1))))
+
+    val df2 = spark.createDataFrame(
+      sparkContext.parallelize(Row(nestedStructValues2) :: Nil),
+      StructType(Seq(StructField("topLevelCol", nestedStructType2))))
+
+    var unionDf = df1.unionByName(df2, true)
+    assert(unionDf.schema.toDDL == "`topLevelCol` " +
+      "STRUCT<`b`: STRUCT<`ba`: STRING, `bb`: STRING>, `a`: STRUCT<`aa`: STRING>>")
+    checkAnswer(unionDf,
+      Row(Row(Row("ba", null), null)) ::
+      Row(Row(Row(null, "bb"), Row("aa"))) :: Nil)
+
+    unionDf = df2.unionByName(df1, true)
+    assert(unionDf.schema.toDDL == "`topLevelCol` STRUCT<`a`: STRUCT<`aa`: STRING>, " +
+      "`b`: STRUCT<`bb`: STRING, `ba`: STRING>>")
+    checkAnswer(unionDf,
+      Row(Row(null, Row(null, "ba"))) ::
+      Row(Row(Row("aa"), Row("bb", null))) :: Nil)
```

Comment:
Can you clarify which step fails? I'm reading this test but can't think of how it can fail without this PR.

Reply:
The union throws an analysis exception because the right side gets corrupted. It's the same example I have in the issue: https://issues.apache.org/jira/browse/SPARK-35290

Reply:
I see. From the JIRA, seems the problem is `b:struct<ba:string:bb:string>` becomes `b:struct<aa:string,bb:string>` due to some reasons. Can we fix the bug surgically without changing how the fields are ordered? Or do we have other reasons to change the behavior?

Reply:
I found one tweak that seemed to fix the issue but made the deeply nested case a lot slower to resolve. I went down this route because it sounded like this would have been the ideal behavior originally but sorting seemed like the only option. And it simplifies things a little bit and uses existing struct merging functionality.

Reply:
Can you open a PR for that tweak? If possible it's better to fix a bug without changing the designed behavior.

Reply:
It was "designed" because it was thought to be the only way. The author said this might be a better approach. I can see if I can remember what that update was, never thoroughly tested it

Reply:
I think sorting is reasonable as well. `df1.unionByName(df2)` can always have the same schema of `df2.unionByName(df1)`

Reply:
It's just weird because it's different behavior than every other part of spark. `withColumn` and `withField` both append to the end. And currently before this PR and if the I applied the tweak `df1.unionByname(df2, True)` would not always have the same schema as `df2.unionByName(df1, True)`. It only sorts nested struct fields, not the top level attributes.

Reply:
I see. So this PR mainly proposes to change the behavior, and fixing a bug is just a bonus.

Reply:
I guess you can look at it that way. It was the most effective way to fix the bug and simplify the code in the process (and not introduce a potential performance issue)

Reply:
I originally planed to work on to get rid of the sorting behavior but have no time on it. I think it is more consistent with top-level union by name behavior. Currently this approach looks much simpler than `withField` approach and can easily get rid of the sorting behavior.